### PR TITLE
[RFC|oneMKL]Enable programmatic versioning

### PR DESCRIPTION
**Pull request by vmalia**
# Motivation
- Customers of oneMKL have requested a way to query the spec-compliance version of the implementation.
- The specification defines versioning but not in a programmatic way.
- Addresses #551 

# Inspiration
- SYCL versioning does not add the `revision` number:
--------
image
--------
- OpenMP specification itself keeps only `MAJOR.MINOR` version, with ongoing progress marked as `Technical Report <number>`. The `_OPENMP` macro provides a release date which maps onto a specification version, like this:
```c
#define _OPENMP 202111    //202111: OpenMP 5.2
```

# Proposal
- `<MAJOR><MINOR><REVISION>` such that MINOR and REVISION have two digits.
- Propose a macro for oneAPI and suggest a naming scheme for component macro names.

### CODE REVIEW

**Comment by rscohn2** [comment]
@uxlfoundation/oneapi-spec-maintainers: Please review this versioning proposal

File: doc/versioning.rst
Comment (mkrainiuk) [inline doc/versioning.rst:93]: According to the proposed format shouldn't  `1.2 rev 1` be `10201` here instead of `10102` since `2` is a minor?

File: doc/versioning.rst
Comment (andrewtbarker) [inline doc/versioning.rst:89]: "appending all digits" to what? You are not proposing a macro/variable so I do not understand where these digits go.

**Review by andrewtbarker — COMMENTED** [review]
I do not understand what is being proposed here.

Is the idea that each implementation of the spec gets to decide how to provide a version to the user (whether by a version string, a `#define`, an `int getSpecVersion()` function)? Then this change to the spec only specifies that the end of this string/macro/int has a certain representation?

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst]: Fixed

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst:89]: Added macro names.

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst:96]: > The PR title suggests that this change only affects oneMKL but it seems to be the scope of this change is broader that than.

I updated the title. You are right, it affects the full document and all components.

> What is the benefit of having ONEAPI_SPEC_VERSION  over only defining \<COMPONENT\>_SPEC_VERSION?

The spec currently does not support versioning on a component level, only oneAPI level, so the `<COMPONENT>_SPEC_VERSION` is optional, but `ONEAPI_SPEC_VERSION` is mandatory.

Comment (vmalia) [reply doc/versioning.rst]: Thanks, done.

Comment (akukanov) [reply doc/versioning.rst]: A mandatory macro with the same name in different projects is a terrible idea. On the user side, it can always happen that libraries implementing a different version of the specification are used, in which case one gets macro redefinition error.

**Comment by vmalia** [comment]
> Is the idea that each implementation of the spec gets to decide how to provide a version to the user (whether by a version string, a `#define`, an `int getSpecVersion()` function)? Then this change to the spec only specifies that the end of this string/macro/int has a certain representation?

Yes. The goal of this RFC is to ensure that every implementation of the spec provides a numerically comparable value for its compliance with the spec version, through the proposed macros.

The implementation can additionally provide more ways to query that information, like APIs, but we do not mandate that in this proposal.

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst:1]: Where should the changes be added then, so that they are reflected in the main oneAPI spec at https://oneapi-spec.uxlfoundation.org/specifications/oneapi/v1.3-rev-1/ ?

File: doc/versioning.rst
Comment (Rbiessy) [reply doc/versioning.rst]: Thanks, I don't have the rights to resolve the comment so feel free to do it when I thumb up.

Comment (Rbiessy) [reply doc/versioning.rst]: Yes I also have concerns about defining the same macro in multiple projects.

File: doc/versioning.rst
Comment (akukanov) [inline doc/versioning.rst:44]: Is it even necessary to track revisions (in other words, editorial changes) in any implementation version macros? My understanding is that any changes that affect API semantics (including observable side effects) require at least the minor version update.

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst]: Updated

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst:44]: The existing text from the spec says:
> Increment the:
> - MAJOR version when adding major new functionality and making incompatible API changes, including removing APIs.
> - MINOR version when adding minor functionality and API changes that are backwards compatible.
> - REVISION when **making backwards compatible bug fixes** or any editing change in the document, including minor changes such as correcting typos. Initial REVISION is 1.

My understanding is that the existing spec allows incrementing revision number for bug fixes, while major and minor updates may contain new functionality. This RFC does not modify this behavior.

If we want to remove the revision from macros, then the revision field should explicitly define only documentation changes, without the **bold text** above. Since this will require a separate RFC, I consider this out-of-scope for this proposal.

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst:1]: @rscohn2 can you please help understand if my changes are in the right place?

File: doc/versioning.rst
Comment (akukanov) [reply doc/versioning.rst:44]: This document describes versioning of the specification, therefore in the context of it "bug fixes" mean changes in the specification, and not in any of the implementations. Therefore, I do not think it says about any new functionality - since bugs fixed in the specification obviously do not involve any source code changes.

As described above, if the specification change includes a new minor functionality or compatible API changes, then the minor version needs to be updated. Therefore I remain positive that new revisions only contain editorial changes but no changes to the API semantics.

I am happy to be corrected, ideally with practical examples. Tagging @rscohn2.

File: doc/versioning.rst
Comment (akukanov) [reply doc/versioning.rst:1]: I would suggest to go to each component library and propose a patch that describes the spec versioning macro for that very library, This will need to be done anyway I believe, even with a common page like you suggest.

The specification source files that are not specific to any particular library reside directly in the `source` folder.

Comment (vmalia) [reply doc/versioning.rst:1]: Are you suggesting that we add the macro name and its description to each component of the oneAPI spec, like oneMKL and oneTBB? What benefit will we get compared to having it in the common versioning section? My understanding is that we would end up duplicating the content, and any future change would need to happen in multiple places.

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst:44]: I am okay to remove the revision part from the proposal if that was the intended use case for the revision field and is currently being followed. I still think the bold text can be improved to have less ambiguity.

File: doc/versioning.rst
Comment (akukanov) [reply doc/versioning.rst:1]: To avoid any ambiguity in the macro names and values, just giving a couple of examples is insufficient. In other words, each library specific macro should be well specified. I believe it is best done in the respective sections, to have all relevant information about a library in a single place. It does not preclude describing some general rules in the "shared" part of the specification and referring to that part where needed.

Comment (vmalia) [reply doc/versioning.rst:1]: > ...each library specific macro should be well specified.

I am not sure I understand what you mean by this. Can you provide some example?

Comment (akukanov) [reply doc/versioning.rst:1]: https://eel.is/c++draft/cpp.predefined
https://www.openmp.org/spec-html/5.1/openmpse10.html
https://www.mpi-forum.org/docs/mpi-4.1/mpi41-report/node244.htm

In all cases, the names of the macros are explicitly written, and the values - or rules for building the values - are unambiguous and clear. Well, for OpenMP it is unknown which combination of a year and a month corresponds to which version; I have not looked carefully if this mapping exists somewhere.

Here comes SYCL: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_preprocessor_directives_and_macros

> The version of SYCL defined in this document will have SYCL_LANGUAGE_VERSION substitute the integer 2020, composed with the general SYCL version followed by 2 digits representing the revision number;

This one I do not clearly understand. Is the value 2020 or 202009 or what? But at least, the name is clear.

The rules in this proposal are somewhat ambiguous for both the name and the value, in my view. Yes, one can take an educated guess - but it will still be a guess.

If a clear definition similar to `__cplusplus` and `MPI_VERSION` needs to be given for each oneAPI specification component, in my opinion the best place for it would be the chapter of that component.

File: doc/versioning.rst
Comment (rscohn2) [reply doc/versioning.rst:44]: In 4 years, I have only used revision for formatting/editing. I can remove the "compatible buf fixes" clause.

File: doc/versioning.rst
Comment (rscohn2) [reply doc/versioning.rst:1]: Yes, the original page probably should have been part of the spec, not part of the documnetation. I don't have any strong opinions about whether this should be common or per element. There would not be a lot of duplication or confusion to have it in the element instead of common so I suggest going with Alexey's desire to move it in MKL.

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst:44]: Thank you @rscohn2 . In that case, I will update the RFC to include only major and minor versions in the macro.

File: doc/versioning.rst
Comment (vmalia) [reply doc/versioning.rst:1]: @akukanov I updated the oneMKL source. Is this better?

Comment (akukanov) [reply doc/versioning.rst:1]: I think it is certainly better than the previous "global" proposal. To me it looks reasonable, but it's now the question for the oneMKL team here. I can think of something similar added for oneDPL and oneTBB.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (spencerpatty) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:22]: I have a concern that over time we may reach a MAJOR version larger than 9, what will happen then?  should we always use 4 digits ?  0101,  0102, etc ?

Comment (spencerpatty) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:22]: actually since we are using a macro number through `#define`, I suppose it really doesn't matter, does it ...

Comment (spencerpatty) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:22]: why are we ignoring the revision ?

**Review by Rbiessy — APPROVED** [review]

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:22]: For the major version, yes, we will keep enabling additional digits as we move beyond `9` or `99`, because it won't affect the numerical comparison. Having a `0` prefix for single digit major versions will need us to update the spec if we ever approach `100`. I did not see a strong reason to do this, so I kept it simple. It is still an option if we feel it will add some advantage.

We are ignoring the revision because as per https://github.com/uxlfoundation/oneAPI-spec/pull/561#discussion_r1725767506, the revision is used only to update the documentation, not the implementation.

**Comment by vmalia** [comment]
@akukanov @mkrainiuk @rscohn2 @spencerpatty if everything looks okay, then please provide your approval. Thank you!

**Review by spencerpatty — APPROVED** [review]
Acceptable from my perspective.  Approved

**Review by mkrainiuk — APPROVED** [review]

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (mkrainiuk) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:43]: Should we mention that `rev 2` is assuming to be the latest one since it's not present in the version macro?

**Comment by vmalia** [comment]
@aelizaro @paveldyakov @lhuot @zettai-reido @sknepper can you please provide your reviews? Thank you!

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:43]: Added a line

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (spencerpatty) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:23]: can we use "SPBLAS" instead of "SPARSE" as someday there will also be SPSOLVERS  sparse solvers and the impls may not be in perfect lock step with SPBLAS ...  but will attempt to be interoperable

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (spencerpatty) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:21]: If/When we switch to oneMATH name for the Specification and Interfaces project, will we also switch the macro names ?  maybe we want to delay this merge until those are made to not have to deprecate/switch ...  ?

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:23]: @spencerpatty To use SPBLAS, we need to define SPBLAS as a domain in the spec's architecture section first, then make the corresponding change here. For now I cannot justify the divergence of the macro name from the domains defined in the spec here: https://oneapi-spec.uxlfoundation.org/specifications/oneapi/v1.3-rev-1/elements/onemkl/source/architecture/architecture#onemkl-namespaces

If an implementation is not compliant with the spec, then any macro name can be used by that implementation. If your suggestion is to make that change, I can create a new issue, but it would be out of scope for this RFC.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:21]: @spencerpatty The changes you mentioned are from https://github.com/oneapi-src/oneMKL/pull/564 which is an RFC in one of the implementations of this spec. The implementation can choose to take any direction, whether or not it is in compliance with the specification.

@Rbiessy you approved this PR earlier so there are 2 questions - 

1. Does your RFC(https://github.com/oneapi-src/oneMKL/pull/564) block this RFC?
2. Does your RFC aim to make changes to the oneAPI specification eventually? If yes, can you please create a placeholder Issue/PR in this project and mention it here? If not, please note that your oneMKL Interfaces RFC breaks compliance with the oneAPI specification as of the current status of your RFC.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (spencerpatty) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:23]: so SPBLAS is a domain and it uses the "sparse" namespace.  Some day in the future there will also be a SPSOLVER  or sparse_solver domain introduced and it will also use the "sparse" namespace ...  I don't see the problem unless you are equating namespace with domain

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:21]: I had an offline discussion with the team and we came to an agreement that the `oneMATH` proposal has a lot of work ahead of it, and that name change will have to be made in not just oneMKL Interfaces, but also in multiple places across oneAPI spec and other documents. Since this change will have a wider impact than just this specific part of document where this PR targets, it is better to make that change at that time. It will not affect the core version schema or approach proposed here.
This PR and RFC is NOT blocked by @Rbiessy 's RFC in oneMKL Interfaces.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:23]: @spencerpatty When we add multiple sparse-related functionality, I think it makes sense to make the namespaces more clear, and if required, distinct for each Sparse group of APIs/algorithms, like SPBLAS, SPSOLVER etc.
I would like to refer to the point 5 of the https://github.com/mkrainiuk/oneMKL/blob/rfcs/README.md#rfc-ratification-process which requires any contributor to propose changes to oneAPI spec first, then make those changes in your backend implementation, if you want to comply with the oneAPI specification.

If you think having SPBLAS is better, then please feel free to open a new GH issue for the same to be reflected in all places in the oneAPI specification. Changing existing terminologies for domains and namespaces is out-of-scope for this RFC.

Comment (spencerpatty) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:23]: namespaces won't be changing, but the name of the domains are different and they would share a namespace ... The name of the domain is not SPARSE but SPBLAS,  in some future time the domain of SPSOLVERS would also be added, but both are separate domains that will share a namespace.  I'm not sure why we need to have a GH issue ...  

https://github.com/uxlfoundation/oneAPI-spec/tree/main/source/elements/oneMKL/source/domains/spblas  the folder name of domain is "spblas" and the namespace is "sparse".  The name of the domain is Sparse BLAS here: 
 https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneMKL/source/domains/spblas/spblas.rst of which "SPBLAS" is short.
 
 On the other side, there is this page about "sparse linear algebra" https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneMKL/source/domains/sparse_linear_algebra.inc.rst which points to spblas.rst as the current only domain in sparse linear algebra, but again this is similar to the "dense linear algebra" page containing BLAS and LAPACK domains in it ...  and at some point sparse solvers will be part of the sparse linear algebra set as well.  
 
 All that being said, to me "SPBLAS" should be considered as the default here, and "SPARSE" would be in my mind deviating from existing documentation and specification and require github issue and PR changes...

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (paveldyakov) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:25]: I can see that Stats domain is missed: https://oneapi-spec.uxlfoundation.org/specifications/oneapi/v1.3-rev-1/elements/onemkl/source/domains/stats/onemkl_stats_overview

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (Rbiessy) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:21]: It does not matter to me which RFC is implemented first. The RFC for renaming is planning to rename the occurrences of all the macros using `ONEMKL`.
The RFC describes how it impacts the specification. I read the ratification process, it does not say that an issue needs to be created. @mkrainiuk could this be clarified in the RFC process? I created https://github.com/uxlfoundation/oneAPI-spec/issues/580 but I have some concerns the discussions will be split in 2 different places now.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:25]: Good catch. I could not find a corresponding namespace.
https://oneapi-spec.uxlfoundation.org/specifications/oneapi/v1.3-rev-1/elements/onemkl/source/architecture/architecture
Is this expected?

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst]: I updated the RFC to reflect SPBLAS as the macro name, and we now have some more work to clarify what we want to do with the sparse functionality in the spec itself: https://github.com/uxlfoundation/oneAPI-spec/issues/581

This hopefully unblocks this proposal from that confusion.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:25]: We should discuss this in https://github.com/uxlfoundation/oneAPI-spec/issues/581 as well.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (mkrainiuk) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:21]: Hi @Rbiessy, good point, we need to clarify it in the ratification process, and I think any changes related to API should be discussed in Spec only, since opensource is just an implementation of the Spec, I believe it should just follow what was decided for Spec. In this particular case I guess the problem is that we have opensource project migration (which is not related to the Spec) combined with renaming (which is Spec related and should be started in the Spec).

Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:21]: Thank you for clarifying, @mkrainiuk . @spencerpatty @Rbiessy Let's proceed with this RFC with the `oneMKL` naming. When the `oneMATH` proposal is made to the oneAPI spec, it can address these macros and other changes across the oneAPI spec simultaneously.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (al42and) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:17]: Do all headers provide this info? E.g.,

- Will `#include "oneapi/mkl.hpp"` provide `ONEMKL_*_SPEC_VERSION` for all domains?
  - If a domain is not supported/installed, will it be undefined of `0`?
- Will `#include "oneapi/mkl/rng.hpp"`  provide `ONEMKL_RNG_SPEC_VERSION`?
- Will `#include "oneapi/mkl/rng.hpp"` provide `ONEMKL_BLAS_SPEC_VERSION`?
- Will `oneapi/mkl/rng/device.hpp` provide `ONEMKL_RNG_SPEC_VERSION`?

Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:17]: @al42and good question. The specification(and this RFC) does not define any implementation-specific header files for oneMKL, so the location of the macro can be decided by the implementation itself.

> If a domain is not supported/installed, will it be undefined or 0?

For an undefined domain implementation: https://oneapi-src.github.io/oneMKL/create_new_backend.html#create-wrappers shows how the wrappers in oneMKL Interfaces work and how the implementation can handle errors for unimplemented backends. Note that this is different than the oneAPI specification that we have here. More clarification on that: https://github.com/oneapi-src/oneMKL/blob/develop/README.md#faqs

I added a rule for the macro numbering for the non-compliance of an implementation. Thanks!

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (paveldyakov) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:25]: Suggestion to add it as "STATS" - such namespace is used for statistics routines

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (paveldyakov) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:41]: Question: Can domain be aligned with provisional spec version? 
E.g. the latest available release is 1.4. provisional https://github.com/uxlfoundation/oneAPI-spec/releases 

If yes, how we should update the macro in this case?

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (sknepper) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:29]: Should we link to a document that describes how/when the MAJOR, MINOR, and revision versions get updated?

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (sknepper) [inline source/elements/oneMKL/source/architecture/other_architecture.inc.rst:17]: The oneMKL spec allows for an implementation of the spec to implement just 1 API and throw unsupported for all of the other APIs in that domain. (See https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneMKL/source/architecture/api_design.inc.rst)
In that case, is the implementation considered conformant to whatever version of the spec has the APIs that are included in the implementation (albeit mostly unsupported), or is that implementation considered non-compliant?

Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:17]: My understanding is that the implementation will still be compliant, because it follows the definitions of the specification to handle errors for unimplemented APIs.

However, the implementation would not be conformant to the standards that the industry would expect backend libraries to be, because of lack of most APIs.
Alternatively, it could be a feature that a specific hardware-software use case only requires or supports certain APIs as per their app's needs.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:25]: done

Comment (paveldyakov) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:25]: Thanks!

**Review by paveldyakov — APPROVED** [review]

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:41]: @paveldyakov the provisional versioning (https://github.com/uxlfoundation/oneAPI-spec/blob/main/doc/versioning.rst#provisional-versions) is for specifications that are pending approval.

It is highly unlikely that we will have implementations that will comply with provisional specifications unless the approval of the spec happens later than the implementation in the backend library. If the backend is ready before the spec, the macro can point to a future non-provisional revision because the spec release follows a predictable convention.

From the above link:
> A specification may be published before approval, but must be labeled provisional. Provisional specifications are published as a series of revisions until approval. After approval, the provisional label is removed and the rev is set to 1.

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:29]: Added at the end of this section.

**Comment by akukanov** [comment]
For your information, below is what I propose for defining a spec version macro for oneDPL. Use as you like; in case of feedback please comment in the PR #591.
```
A oneDPL implementation must define a preprocessor macro representing the version
of the oneDPL specification that the implementation is compliant with.

.. code:: cpp
  // Defined in <oneapi/dpl/version>
  #define ONEDPL_SPEC_VERSION /*implementation-defined*/

The ``ONEDPL_SPEC_VERSION`` macro must be defined to the decimal literal which value equals to
*major-spec-version* * 100 + *minor-spec-version*, where *major-spec-version* and *minor-spec-version*
are the major and the minor versions of the latest fully supported specification.
[*Example*: ``#define ONEDPL_SPEC_VERSION 104`` for the oneDPL specification 1.4.  -- *end example*]
```

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (al42and) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:17]: > For an undefined domain implementation, this, oneapi-src.github.io/oneMKL/create_new_backend.html#create-wrappers shows how the wrappers in oneMKL Interfaces work and how the implementation can handle errors for unimplemented backends.

I was referring not to the case of unimplemented _backends_, but to the case of a _domain_ missing entirely: "It is permitted for a library implementation of the oneMKL specification to implement either all, one or more than one of the domains in oneMKL."

But now that I read it, the wording "Each oneMKL domain must define a preprocessor macro [...]" is quite clear that if domain is not supported, the macro is not defined.

Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:17]: @al42and yes, you are right.

**Comment by vmalia** [comment]
@rscohn2 we have all the required feedback. Can you please help to merge this PR?

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (paveldyakov) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:41]: sounds good, thanks!

File: source/elements/oneMKL/source/architecture/other_architecture.inc.rst
Comment (vmalia) [reply source/elements/oneMKL/source/architecture/other_architecture.inc.rst:21]: The name transition and corresponding changes from this proposal are part of https://github.com/uxlfoundation/oneAPI-spec/pull/596 now.

**Pull request merged**
